src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2996,26 +2996,69 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
                 {
-                    if (stmt.Kind != BoundKind.LocalFunctionStatement)
+                    if (stmt is BoundLocalFunctionStatement localFunc)
+                    {
+                        localFuncs.Add(localFunc);
+                    }
+                    else
                     {
                         VisitStatement(stmt);
                     }
                 }
 
-                // Now visit the local function bodies
-                foreach (var stmt in block.Statements)
+                // Now visit the local function bodies.
+                // We first visit those we have seen called (hence we know their starting state),
+                // repeating this while visiting new bodies (which might contain more call sites).
+                // This avoids unnecessary passes and incorrect starting states of reachable bodies
+                // (which could be captured in loop head state for example).
+                for (bool newBodiesVisited = true; newBodiesVisited;)
                 {
-                    if (stmt is BoundLocalFunctionStatement localFunc)
+                    newBodiesVisited = false;
+
+                    // We visit functions in reverse order, so deleting from the list has good perf in the common case.
+                    for (int i = localFuncs.Count - 1; i >= 0; i--)
+                    {
+                        var localFunc = localFuncs[i];
+
+                        // We visit the body only if the function's usages state has been created
+                        // which happens when we visit the function's call site or its body.
+                        // In the first pass of the nullable walker, existence of usages here means that a call site has been visited.
+                        // In subsequent nullable walker passes, the starting state is preserved from previous passes.
+                        // In any case, existence of usages means that we have a good starting state.
+                        if (HasLocalFuncUsagesCreated(localFunc.Symbol))
+                        {
+                            localFuncs.RemoveAt(i);
+
+                            // The body of this local function might contain calls to other local functions,
+                            // hence we will rerun the outer loop to visit bodies of newly-called functions if any.
+                            newBodiesVisited = true;
+
+                            TakeIncrementalSnapshot(localFunc);
+                            VisitLocalFunctionStatement(localFunc);
+                        }
+                    }
+
+                    // If we haven't visited new bodies in this iteration, visit an unreachable function if any.
+                    // This might make other functions reachable, so we will continue with the outer loop.
+                    if (!newBodiesVisited && localFuncs.Count != 0)
                     {
+                        var localFunc = localFuncs[^1];
```

Comment:
>var localFunc = localFuncs[^1]; [](http://example.com/codeflow?start=24&length=31)

Here, I think we should grab the lexically first "unused" local function instead. Even if that means that we need to copy more items during removal. I do not think it would be common to have many completely unreachable local functions at the same time, therefore, the amount of items to shift won't be too much to cause a problem

Reply:
Alternatively, we could avoid doing any remove operations in the ```localFuncs``` list. We could simply set handled items to null and skip over nulls during iteration. Keeping track of an "active" count in a dedicated local might help with quick count checks as well.